lib/checkstl.cpp

```diff
@@ -1869,3 +1869,268 @@ void CheckStl::readingEmptyStlContainerError(const Token *tok, const ValueFlow::
+    input = Token::findmatch(assignTok->next(), "%varid%", endStatement, varid);
+    return assignTok;        
+}
+
+static const Token* singleMemberCallInScope(const Token* start, unsigned int varid, bool& input)
+{
+    if(start->str() != "{")
+        return nullptr;
+    const Token * endToken = start->link();
+    if(!Token::Match(start->next(), "%var% . %name% ("))
+        return nullptr;
+    if(!Token::Match(start->tokAt(4)->link(), ") ; }"))
+        return nullptr;
+    const Token * endStatement = start->tokAt(4)->link()->next();
+    if(endStatement->next() != endToken)
+        return nullptr;
+
+    const Token * dotTok = start->tokAt(2);
+    if(!Token::findmatch(dotTok->tokAt(2), "%varid%", endStatement, varid))
+        return nullptr;
+    input = Token::Match(start->next(), "%var% . %name% ( %varid% )", varid);
+    if(isVariableChanged(dotTok->next(), endStatement, dotTok->astOperand1()->varId(), false, nullptr, true))
+        return nullptr;
+    return dotTok;      
+}
+
+static const Token* singleIncrementInScope(const Token* start, unsigned int varid, bool& input)
+{
+    if(start->str() != "{")
+        return nullptr;
+    const Token * varTok = nullptr;
+    if(Token::Match(start->next(), "++ %var% ; }"))
+        varTok = start->tokAt(2);
+    else if(Token::Match(start->next(), "%var% ++ ; }"))
+        varTok = start->tokAt(1);    
+    if(!varTok)
+        return nullptr;
+    input = varTok->varId() == varid;
+    return varTok;      
+}
+
+static const Token* singleConditionalInScope(const Token* start, unsigned int varid)
+{
+    if(start->str() != "{")
+        return nullptr;
+    const Token * endToken = start->link();
+    if(!Token::Match(start->next(), "if ("))
+        return nullptr;
+    if(!Token::Match(start->tokAt(2)->link(), ") {"))
+        return nullptr;
+    const Token * bodyTok = start->tokAt(2)->link()->next();
+    const Token * endBodyTok = bodyTok->link();
+    if(!Token::Match(endBodyTok, "} }"))
+        return nullptr;
+    if(endBodyTok->next() != endToken)
+        return nullptr;
+    if(!Token::findmatch(start, "%varid%", bodyTok, varid))
+        return nullptr;
+    if(isVariableChanged(start, bodyTok, varid, false, nullptr, true))
+        return nullptr;
+    return bodyTok;
+}
+
+static bool addByOne(const Token * tok, unsigned int varid)
+{
+    if(Token::Match(tok, "+= %any% ;") && 
+        tok->tokAt(1)->hasKnownIntValue() && 
+        tok->tokAt(1)->getValue(1)) {
+        return true;
+    }
+    if(Token::Match(tok, "= %varid% + %any% ;", varid) && 
+        tok->tokAt(3)->hasKnownIntValue() && 
+        tok->tokAt(3)->getValue(1)) {
+        return true;
+    }
+    return false;
+}
+
+static bool accumulateBool(const Token * tok, unsigned int varid)
+{
+    if(Token::Match(tok, "=|&=|%or%= %bool% ;") && 
+        tok->tokAt(1)->hasKnownIntValue()) {
+        return true;
+    }
+    if(Token::Match(tok, "= %varid% &|%oror%|%or%|&& %bool% ;", varid) && 
```

Comment:
in the beginning , the `%..%` patterns had to come first. Today it does not really matter. But I would still prefer `%oror%|%or%|&&|&`. This knowledge could be used in `Token::Match` to speed it up a little, I think.